### PR TITLE
TCP/IPv4 adaptor for CoAP

### PR DESCRIPTION
**Pull request by ccollins476ad**
This transport uses mn_sockets.

Unlike other transports, this one does not listen for new connections on its own.  Instead, the application needs to populate the transport with connections to be managed using oc_tcp4_add_conn().  There are a few too many parameters that need to be specified when listening for TCP connections.

### CODE REVIEW

**Comment by vrahane** [comment]
I guess we can call datagram transport as UDP but well it won't be a correct term for BLE :-)

**Comment by vrahane** [comment]
@ccollins476ad I think there is a style check issue most probably, hence the CI failure, rest looks good.

File: net/oic/src/messaging/coap/engine.c [hunk 1]
```diff
@@ -231,17 +231,19 @@ coap_receive(struct os_mbuf **mp)
...
+            /* handle responses */
+            if (message->type == COAP_TYPE_CON) {
+                erbium_status_code = EMPTY_ACK_RESPONSE;
+            } else if (message->type == COAP_TYPE_ACK) {
+                /* transactions are closed through lookup below */
+            } else if (message->type == COAP_TYPE_RST) {
 #ifdef OC_SERVER
-            /* cancel possible subscriptions */
-            coap_remove_observer_by_mid(OC_MBUF_ENDPOINT(m), message->mid);
+                /* cancel possible subscriptions */
```
Comment (vrahane) [inline net/oic/src/messaging/coap/engine.c:243]: more spaces than required

Comment (ccollins476ad) [reply net/oic/src/messaging/coap/engine.c:243]: I think it is correct.  The `#ifdef` just makes it look confusing.

```
        if (!oc_endpoint_use_tcp(&endpoint)) {
            /* handle responses */
            if (message->type == COAP_TYPE_CON) {
                erbium_status_code = EMPTY_ACK_RESPONSE;
            } else if (message->type == COAP_TYPE_ACK) {
                /* transactions are closed through lookup below */
            } else if (message->type == COAP_TYPE_RST) {
#ifdef OC_SERVER
                /* cancel possible subscriptions */
                coap_remove_observer_by_mid(OC_MBUF_ENDPOINT(m), message->mid);
#endif
            }
```

File: net/oic/include/oic/port/mynewt/stream.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Indicates whether a transport-specific endpoint matches the provided
+ * description.
+ */
+typedef bool oc_stream_ep_match(const void *ep, void *ep_desc);
```
Comment (kasjer) [inline net/oic/include/oic/port/mynewt/stream.h:40]: Any reason why only ep is const?

File: net/oic/src/port/mynewt/stream.c [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#include "oic/messaging/coap/coap.h"
+
+int
+oc_stream_reass(struct oc_stream_reassembler *r, struct os_mbuf *om1,
+                void *ep_desc, struct os_mbuf **out_pkt)
+{
+    struct os_mbuf_pkthdr *pkt1;
+    struct os_mbuf_pkthdr *pkt2;
+    struct os_mbuf *om2;
+    uint8_t hdr[sizeof (struct coap_tcp_hdr32)];
```
Comment (kasjer) [inline net/oic/src/port/mynewt/stream.c:36]: This is the only place where space is after sizeof

File: net/oic/src/port/mynewt/tcp4_adaptor.c [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+ * under the License.
+ */
+
+#include <assert.h>
+#include <string.h>
+#include <stdint.h>
+
+#include "os/mynewt.h"
+
+uint8_t oc_tcp4_transport_id = -1;
```
Comment (kasjer) [inline net/oic/src/port/mynewt/tcp4_adaptor.c:26]: Slightly unorthodox way to initialize unsigned value.
Other place in this commit use UINT8_MAX

File: net/oic/src/port/mynewt/tcp4_adaptor.c [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+    struct mn_socket *sock;
+    oc_tcp4_err_fn *err_cb;
+    void *err_arg;
+    SLIST_ENTRY(oc_tcp4_conn) next;
+};
+
+struct os_mempool oc_tcp4_conn_pool;
+static os_membuf_t oc_tcp4_conn_buf[
+    OS_MEMPOOL_SIZE(MYNEWT_VAL(OC_TCP4_MAX_CONNS),
+                    sizeof (struct oc_tcp4_conn))
```
Comment (kasjer) [inline net/oic/src/port/mynewt/tcp4_adaptor.c:120]: Second place with space after sizeof few lines above there was no space

File: net/oic/include/oic/port/mynewt/stream.h (hunk 2 above)
Comment (ccollins476ad) [reply net/oic/include/oic/port/mynewt/stream.h]: Thanks.  I agree `ep_desc` should also be const here.  I also think it should be const in the "fill" callback.  I have fixed both.

File: net/oic/src/port/mynewt/stream.c (hunk 3 above)
Comment (ccollins476ad) [reply net/oic/src/port/mynewt/stream.c]: Thanks, I have fixed this (and the other style issues you identified).

**Pull request merged**